File: src/day4.py
offsets = [(0, 1), (-1, -1), (-1, 1), (1, 1), (1, -1), (-1, 0), (1, 0), (0, -1)]



def remove(lines) -> int:
    removed = 0

    for r, line in enumerate(lines):
        for c, val in enumerate(line):
            if val != "@":
                # print(".", end="")
                continue

            adjacent_count = 0

            for dr, dc in offsets:
                nr, nc = r + dr, c + dc
                if nr < 0 or nc < 0 or nr >= len(lines) or nc >= len(line):
                    continue

                adjacent = lines[nr][nc]

                if adjacent == "@":
                    # print("@")
                    adjacent_count += 1

            if adjacent_count < 4:
                lines[r][c] = "."
                removed += 1
    
    return removed


with open('data/day4.txt', 'r') as f:
    lines = f.read().splitlines()


lines = [list(line) for line in lines]
print(lines[0])

removed = 0

while True:
    new_removed = remove(lines)
    if new_removed == 0:
        break

    removed += new_removed

print(removed)



            # print(adjacent_count, end="")
            # print("x", end="")
        # else:
            
            # print(adjacent_count, end="")
            # print("@", end="")
    

print(removed)

File: src/day4.rs
fn part1(rows: Vec<Vec<u8>>) -> i32 {
    let offsets: Vec<(i32, i32)> = vec![
        (-1, 1),
        (1, 0),
        (1, 1),
        (-1, 0),
        (1, 0),
        (-1, -1),
        (-1, 0),
        (1, -1),
    ];

    let mut count = 0;

    for i in 0..rows.len() {
        for j in 0..rows.get(i).unwrap().len() {
            if rows.get(i).unwrap().get(j).unwrap() != &b'@' {
                continue
            } 

            let mut adjacent = 0;
            for (di, dj) in offsets.iter() {
                // ??????????????????????
                if let Some(new_i) = i.checked_add_signed(*di as isize) {
                    if let Some(new_j) = j.checked_add_signed(*dj as isize) {
                        
                        if let Some(row) = rows.get(new_i) {
                            if let Some(chr) = row.get(new_j) {
                                if chr == &b'@' {
                                    adjacent += 1;
                                }
                            }
                        }
                    }
                }
            }

            if adjacent < 4 {
                count += 1;
            }
        }
    }

    count
}

fn main() {
    let path = std::env::args()
        .nth(1)
        .unwrap_or_else(|| "data/day4.txt".to_string());

    println!("reading {}", path);

    let data = std::fs::read_to_string(path).unwrap();

    let rows: Vec<Vec<u8>> = data.lines().map(|line| line.bytes().collect()).collect();

    println!("{}", part1(rows));
}
